src/SourceGenerators/InstrumentationDefinitionsGenerator.cs

```diff
@@ -0,0 +1,280 @@
+// <copyright file="InstrumentationDefinitionsGenerator.cs" company="OpenTelemetry Authors">
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+// </copyright>
+
+using System.Collections.Immutable;
+using System.Text;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Text;
+
+namespace SourceGenerators;
+
+[Generator]
+public class InstrumentationDefinitionsGenerator : IIncrementalGenerator
+{
+    public void Initialize(IncrementalGeneratorInitializationContext context)
+    {
+        var classesWithAttributes = context.SyntaxProvider.CreateSyntaxProvider(
+                static (node, _) => IsAttributedClass(node),
+                static (node, _) => GetClassesMarkedByInstrumentMethodAttribute(node))
+            .Where(static m => m != null);
+
+        IncrementalValueProvider<(Compilation, ImmutableArray<ClassDeclarationSyntax?>)> compilationAndClasses
+            = context.CompilationProvider.Combine(classesWithAttributes.Collect());
+
+        context.RegisterSourceOutput(
+            compilationAndClasses,
+            static (context, source) => Generate(source.Item1, source.Item2, context));
+    }
+
+    private static void Generate(Compilation compilation, ImmutableArray<ClassDeclarationSyntax?> classesWithAttributes, SourceProductionContext context)
```

Comment:
If my understanding is correct:
```suggestion
    private static void Generate(Compilation compilation, ImmutableArray<ClassDeclarationSyntax?> instrumentationClasses, SourceProductionContext context)
```